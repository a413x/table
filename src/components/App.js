import React, { Component } from "react";

import Search from './Search.js';
import Table from './Table.js';
import PagesNav from './PagesNav.js';
import SelectedUser from './SelectedUser.js';
import AddForm from './AddForm.js';
import LoadData from './LoadData.js';

//Колонки таблицы
//field - ключ данных в базе данных,
//title - имя колонки в шапке таблицы,
//type - тип данных для валидации при добавлении строки
const tableColumns = [{field:'id',title:'#',type:'number'},
  {field:'firstName',title:'First name',type:'text'},
  {field:'lastName',title:'Last name',type:'text'},
  {field:'email',title:'E-mail',type:'email'},
  {field:'phone',title:'Phone',type:'phone'}];

class App extends Component {
    constructor(props){
      super(props);
      this.state = {
        users:[],
        page:1,
        numberOfShownUsers:5,
        selectedUser:null,
        isLoading: false
      };
      this.usersCopy = [];
      this.searchFunc = this.searchFunc.bind(this);
      this.onNumberShownChange = this.onNumberShownChange.bind(this);
      this.getShownData = this.getShownData.bind(this);
      this.onPageChange = this.onPageChange.bind(this);
      this.sortFunc = this.sortFunc.bind(this);
      this.onRowClick = this.onRowClick.bind(this);
      this.addUser = this.addUser.bind(this);
      this.onLoadData = this.onLoadData.bind(this);
    }

    async onLoadData(url){
      this.setState({isLoading: true});
      const response = await fetch(url)
        .then(response => response.json())
        .then(data => {
          this.setState({
            users: data,
            isLoading: false
          });
          this.usersCopy = data;
        }).catch(error => {
          alert('Не удалось загрузить данные');
          this.setState({isLoading: false});
        });
    }

    getShownData(page){
      let from = 0;
      const users = this.state.users;
      const numberOfUsers = users.length;
      const numberOfShownUsers = this.state.numberOfShownUsers;
      if(page) from = (page-1)*numberOfShownUsers;
      return users.slice(from,from+numberOfShownUsers);
    }

    searchFunc(str,columnsToSearch){
      const filteredUsers = this.usersCopy.filter((user)=>{
        for(let i = 0; i < columnsToSearch.length; i++){
          const val = user[columnsToSearch[i].field];
          if(typeof val === 'object' || typeof val === 'boolean'
            || typeof val === 'function' || typeof val === 'undefined') continue;
          if((''+val).toLowerCase().includes(str.toLowerCase())) return true;
        }
        return false;
      });
      this.setState({
        users: filteredUsers,
        page: 1
      });
    }

    sortFunc(field,order){
      const sign = order == 'asc' ? 1 : -1;
      const sortedUsers = [...this.state.users];
      sortedUsers.sort((a,b) => {
        let val1 = a[field];
        let val2 = b[field];
        if(typeof val1 === 'string' && typeof val2 === 'string'){
          val1 = val1.toLowerCase();
          val2 = val2.toLowerCase();
        }
        return val1 > val2 ? sign : -sign;
      });
      this.setState({
        users: sortedUsers
      });
    }

    onPageChange(newPage){
      this.setState({
        page:newPage
      });
    }
    onNumberShownChange(val){
      this.setState({
        page:1,
        numberOfShownUsers: val
      });
    }

    onRowClick(user){
      this.setState({selectedUser:user});
    }

    addUser(newUser){
      let newUsers = [...this.state.users];
      newUsers.unshift(newUser);
      this.setState({
        users: newUsers
      });
      this.usersCopy = newUsers;
    }

    render() {
        const currentPage = this.state.page;
        const shownUsers = this.getShownData(currentPage);
        const numberOfPages = Math.ceil(this.state.users.length/this.state.numberOfShownUsers);
        return (
          <div className = 'container'>
            <LoadData onLoadData = {this.onLoadData} isLoading = {this.state.isLoading}/>
            <AddForm columns = {tableColumns} addUser = {this.addUser}/>
            <Search searchFunc = {this.searchFunc} columns = {tableColumns} />
            <Table data = {shownUsers}
              columns = {tableColumns}
              sortFunc = {this.sortFunc}
              onRowClick = {this.onRowClick}/>
            <PagesNav pages = {numberOfPages} currentPage = {currentPage}
              onPageChange={this.onPageChange}
              onNumberShownChange = {this.onNumberShownChange}/>
            {this.state.selectedUser &&
              <SelectedUser user = {this.state.selectedUser}
                unselectRow = {()=>{this.setState({selectedUser:null})}}/>
            }
          </div>
        );
    }
}

export default App;
