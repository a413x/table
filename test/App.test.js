import React from "react";
import {shallow,mount} from 'enzyme';

import App from "../src/components/App";
import data from './data.js';

describe('tests for App component',() => {
  const appContainer = shallow(<App />);
  const instance = appContainer.instance();

  const initialState = {
    users:[],
    page:1,
    numberOfShownUsers:5,
    selectedUser:null,
    isLoading: false
  };

  const fakeData = data.fakeData;
  const tableColumns = data.tableColumns;

  it('renders properly',()=>{
    expect(appContainer).toMatchSnapshot();
  });

  it('initialize with initial state',()=>{
    expect(appContainer.state()).toEqual(initialState);
  });

  it('onLoadData function call',async ()=>{
    jest.spyOn(instance,'onLoadData');

    global.fetch = jest.fn().mockImplementation(() =>
      Promise.resolve({
        json: () => Promise.resolve(fakeData)
      })
    );

    await instance.onLoadData('/');
    expect(instance.onLoadData).toHaveBeenCalledWith('/');
    expect(appContainer.state().users).toEqual(fakeData);
    expect(instance.usersCopy).toEqual(fakeData);
    expect(appContainer.state().isLoading).toBe(false);

    global.fetch.mockClear();
    delete global.fetch;
  });

  it('getShownData function call',()=>{
    jest.spyOn(instance,'getShownData');
    const numberOfUsers = 5;
    appContainer.setState({users:fakeData,numberOfShownUsers:numberOfUsers});
    const shownData = instance.getShownData(0);
    expect(instance.getShownData).toHaveBeenCalledWith(0);
    expect(shownData).toEqual(fakeData.slice(0,numberOfUsers));
  });

  describe('searchFunc function call',()=>{
    jest.spyOn(instance,'searchFunc');
    instance.usersCopy = fakeData;
    let searchStr;
    it('call searchFunc with "2", should find 1 row, users should equal fakeData[1]',()=>{
      searchStr = '2';
      instance.searchFunc(searchStr,tableColumns);
      expect(instance.searchFunc).toHaveBeenCalledWith('2',tableColumns);
      expect(appContainer.state().users).toEqual(fakeData.slice(1,2));
    });
    it('call searchFunc with "d", should find 2 rows, users should include fakeData[1] and fakeData[3]',()=>{
      searchStr = 'd';
      instance.searchFunc(searchStr,tableColumns);
      expect(instance.searchFunc).toHaveBeenCalledWith('d',tableColumns);
      expect(appContainer.state().users).toEqual(fakeData.slice(1,2).concat(fakeData.slice(3,4)));
    });
    it('call searchFunc with "z", should find 0 rows, users should equal []',()=>{
      searchStr = 'z';
      instance.searchFunc(searchStr,tableColumns);
      expect(instance.searchFunc).toHaveBeenCalledWith('z',tableColumns);
      expect(appContainer.state().users).toEqual([]);
    });
  });

  describe('sortFunc function call',()=>{
    jest.spyOn(instance,'sortFunc');
    let field;
    let order;
    it('sort users descending by id',()=>{
      appContainer.setState({users:fakeData});
      field = 'id';
      order = 'desc';
      const sortedData = [...fakeData].sort((a,b) => a[field] > b[field] ? -1 : 1)
      instance.sortFunc(field,order);
      expect(instance.sortFunc).toHaveBeenCalledWith(field,order);
      expect(appContainer.state().users).toEqual(sortedData);
    });
    it('sort users ascending by email',()=>{
      field = 'email';
      order = 'asc';
      const sortedData = [...fakeData].sort((a,b) => a[field] > b[field] ? 1 : -1)
      instance.sortFunc(field,order);
      expect(instance.sortFunc).toHaveBeenCalledWith(field,order);
      expect(appContainer.state().users).toEqual(sortedData);
    });
  });

  it('onPageChange function call',()=>{
    jest.spyOn(instance,'onPageChange');
    const page = 2;
    instance.onPageChange(page);
    expect(instance.onPageChange).toHaveBeenCalledWith(page);
    expect(appContainer.state().page).toBe(page);
  });

  it('onNumberShownChange function call',()=>{
    jest.spyOn(instance,'onNumberShownChange');
    const num = 10;
    instance.onNumberShownChange(num);
    expect(instance.onNumberShownChange).toHaveBeenCalledWith(num);
    expect(appContainer.state().page).toBe(1);
    expect(appContainer.state().numberOfShownUsers).toBe(num);
  });

  it('onRowClick function call',()=>{
    jest.spyOn(instance,'onRowClick');
    appContainer.setState({users:fakeData});
    const user = fakeData[3];
    expect(appContainer.find('SelectedUser')).toHaveLength(0);
    instance.onRowClick(user);
    expect(instance.onRowClick).toHaveBeenCalledWith(user);
    expect(appContainer.state().selectedUser).toEqual(user);
    expect(appContainer.find('SelectedUser')).toHaveLength(1);
  });

  it('addUser function call',()=>{
    jest.spyOn(instance,'addUser');
    const newUser = {id:7,firstName:'Z',lastName:'Z',email:'z@zz.zz',phone:'71111111111'};
    const newUsersArr = [...fakeData];
    newUsersArr.unshift(newUser);
    instance.addUser(newUser);
    expect(instance.addUser).toHaveBeenCalledWith(newUser);
    expect(appContainer.state().users).toEqual(newUsersArr);
  });

  afterAll(() => {
    appContainer.setState(initialState);
  });
});
