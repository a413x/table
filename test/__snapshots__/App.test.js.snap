// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests for App component renders properly 1`] = `
<div
  className="container"
>
  <LoadData
    isLoading={false}
    onLoadData={[Function]}
  />
  <AddForm
    addUser={[Function]}
    columns={
      Array [
        Object {
          "field": "id",
          "title": "#",
          "type": "number",
        },
        Object {
          "field": "firstName",
          "title": "First name",
          "type": "text",
        },
        Object {
          "field": "lastName",
          "title": "Last name",
          "type": "text",
        },
        Object {
          "field": "email",
          "title": "E-mail",
          "type": "email",
        },
        Object {
          "field": "phone",
          "title": "Phone",
          "type": "phone",
        },
      ]
    }
  />
  <Search
    columns={
      Array [
        Object {
          "field": "id",
          "title": "#",
          "type": "number",
        },
        Object {
          "field": "firstName",
          "title": "First name",
          "type": "text",
        },
        Object {
          "field": "lastName",
          "title": "Last name",
          "type": "text",
        },
        Object {
          "field": "email",
          "title": "E-mail",
          "type": "email",
        },
        Object {
          "field": "phone",
          "title": "Phone",
          "type": "phone",
        },
      ]
    }
    searchFunc={[Function]}
  />
  <Table
    columns={
      Array [
        Object {
          "field": "id",
          "title": "#",
          "type": "number",
        },
        Object {
          "field": "firstName",
          "title": "First name",
          "type": "text",
        },
        Object {
          "field": "lastName",
          "title": "Last name",
          "type": "text",
        },
        Object {
          "field": "email",
          "title": "E-mail",
          "type": "email",
        },
        Object {
          "field": "phone",
          "title": "Phone",
          "type": "phone",
        },
      ]
    }
    data={Array []}
    onRowClick={[Function]}
    sortFunc={[Function]}
  />
  <PagesNav
    currentPage={1}
    onNumberShownChange={[Function]}
    onPageChange={[Function]}
    pages={0}
  />
</div>
`;
